### ISSUE TITLE
Deeply nested MULTISET produce quoted JSON in MariaDB

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#12168]
The query introduced as a test for #12155:

```java
Record2<
    Integer,
    Result<Record2<
        Integer,
        Result<Record2<
            Integer,
            Result<Record2<
                Integer,
                Result<Record1<Integer>>
            >>
        >>
    >>
> record =
ctx.select(
        val(1).as("a"),
        multiset(select(
            val(2).as("b"),
            multiset(select(
                val(3).as("c"),
                multiset(select(
                    val(4).as("d"),
                    multiset(select(val(5).as("e")))
                ))
            ))
        ))
    )
   .fetchOne();
```

Looks like this in MySQL and MariaDB:

```sql
select
  1 as `a`,
  (
    select coalesce(
      json_merge_preserve('[]', concat(
        '[',
        group_concat(json_array(`v0`, `v1`) separator ','),
        ']'
      )),
      json_array()
    )
    from (
      select
        2 as `v0`,
        (
          select coalesce(
            json_merge_preserve('[]', concat(
              '[',
              group_concat(json_array(`v0`, `v1`) separator ','),
              ']'
            )),
            json_array()
          )
          from (
            select
              3 as `v0`,
              (
                select coalesce(
                  json_merge_preserve('[]', concat(
                    '[',
                    group_concat(json_array(`v0`, `v1`) separator ','),
                    ']'
                  )),
                  json_array()
                )
                from (
                  select
                    4 as `v0`,
                    (
                      select coalesce(
                        json_merge_preserve('[]', concat(
                          '[',
                          group_concat(json_array(`v0`) separator ','),
                          ']'
                        )),
                        json_array()
                      )
                      from (
                        select 5 as `v0`
                      ) as `t`
                    ) as `v1`
                ) as `t`
              ) as `v1`
          ) as `t`
        ) as `v1`
    ) as `t`
  );
```

It produces the expected result in MySQL:

```
|a  |x                         |
|---|--------------------------|
|1  |[[2, [[3, [[4, [[5]]]]]]]]|
```

But a badly quoted one in MariaDB:

```
|a  |x                                       |
|---|----------------------------------------|
|1  |[[2, "[[3, \"[[4, \\\"[[5]]\\\"]]\"]]"]]|
```

See also:

- https://jira.mariadb.org/browse/MDEV-13701
- https://jira.mariadb.org/browse/MDEV-21902

We probably have to work around this MariaDB bug somehow

**Comment by lukaseder** [comment]
Another batch of `JSON_MERGE_PRESERVE` calls seems to do the trick here:

```sql
select
  1 as `a`,
  (
    select coalesce(
      json_merge_preserve('[]', concat(
        '[',
        group_concat(json_array(`v0`, json_merge_preserve('[]', `v1`)) separator ','),
        ']'
      )),
      json_array()
    ) x
    from (
      select
        2 as `v0`,
        (
          select coalesce(
            json_merge_preserve('[]', concat(
              '[',
              group_concat(json_array(`v0`, json_merge_preserve('[]', `v1`)) separator ','),
              ']'
            )),
            json_array()
          ) x
          from (
            select
              3 as `v0`,
              (
                select coalesce(
                  json_merge_preserve('[]', concat(
                    '[',
                    group_concat(json_array(`v0`, json_merge_preserve('[]', `v1`)) separator ','),
                    ']'
                  )),
                  json_array()
                ) x
                from (
                  select
                    4 as `v0`,
                    (
                      select coalesce(
                        json_merge_preserve('[]', concat(
                          '[',
                          group_concat(json_array(`v0`) separator ','),
                          ']'
                        )),
                        json_array()
                      ) x
                      from (
                        select 5 as `v0`
                      ) as `t`
                    ) as `v1`
                ) as `t`
              ) as `v1`
          ) as `t`
        ) as `v1`
    ) as `t`
  ) x;
```

I've reported this as a bug to MariaDB: https://jira.mariadb.org/browse/MDEV-26134

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.16.0 and 3.15.1 (https://github.com/jOOQ/jOOQ/issues/12170)
